src/main/include/log4cxx/spi/location/locationinfo.h

```diff
@@ -132,9 +152,28 @@ class LOG4CXX_EXPORT LocationInfo
 #if !defined(__LOG4CXX_FUNC__)
 	#define __LOG4CXX_FUNC__ ""
 #endif
+
+#if __cpp_lib_string_view || (_MSVC_LANG >= 201703L)
+#include <string_view>
+
+#define LOG4CXX_LOCATION_CREATE ::std::string_view file_name{__FILE__};\
```

Comment:
I don't get the point of this macro: Isn't this calculated at runtime as well, only differently than `getShortFileName`? Why not have two different calculations of `getShortFileName` in the class instead of a macro containing complex implementation, making debugging more difficult in case of problems? Is there some `constexpr` or `static` missing to calculate only once?

Reply:
Using `string_view` lets it be calculated at compile-time(assuming a high enough optimization level).  Upon some further investigation though, both GCC and MSVC can evaluate `strlen` on a static string at compile time.  GCC at least is also smart enough to evaluate `strrchr` at compile time, so I think we can do this with only C functions.

Basically the following will work in both MSVC and GCC(at least according to godbolt), with GCC managing to figure out at compile-time that 'name' is '/hithere.cpp':
```
const char* file_name = "/bar/foo/hithere.cpp";
const int file_length = strlen(file_name);
const char* name = strrchr(file_name, '/');
```

Unfortunately it seems that with MSVC using either `string_view` or `strrchr` still requires some processing at run-time in order to calculate the correct position.

Reply:
So while the macro copies code in the end multiple times, with a supporting compiler the calculation might happen at compile time because of an optimized `string_view`. But that is not guaranteed and the macro might as well be executed multiple times at runtime?

That seems unpredictable and unnecessary confusing to me. Why not use a static function based on `constexpr` instead, which is C++11 and things either reliably break or work? If `string_view` is supported at compile time, shouldn't that be available with `constexpr` then as well? But that would be guaranteed to happen at compile time.

OTOH, your custom implementation might work with `constexpr` as well which is documented to have support for local [variables and loops since C++14](https://en.cppreference.com/w/cpp/language/constexpr). So I suggest something like the following:

```c++
#define LOG4CXX_LOCATION_CREATE ::log4cxx::spi::LocationInfo location(__FILE__,         \
	::log4cxx::spi::LocationInfo::calcShortFileName(__FILE__), \
	__LOG4CXX_FUNC__, \
	__LINE__)
```

The implementation of that function checks if `string_view` is available and uses that or your custom implementation. Though, with your custom implementation in backwards, `string_view` is redundant anyway. The functions documents to be calculated at compile time and the macro is less verbose as well.

Reply:
Yes, I like your idea of a separate function, that makes it much cleaner.  I was trying to avoid expanding __FILE__ multiple times, until I remembered that the compiler will only put in the same string once into the binary.  That also makes it easier to support both versions: one using `string_view` and the fallback using `strrchr`.